tools/bbbsupervisor/bbbsupervisor.go

```diff
@@ -105,6 +126,64 @@ func (lf logFollower) follow() {
 	errWriter.Write([]byte(fmt.Sprintf("command %v unexpectedly exited", fullCmd)))
 }
 
+// follow implements follower interface by following Prometheus server and querying for values forever.
+func (pl prometheusFollower) follow() {
+	for {
+		fmt.Printf("prometheusFollower querying for %q\n", pl.expression)
+		pl.query()
+		time.Sleep(30000 * time.Millisecond)
+	}
+}
+
+// query queries the Prometheus server once.
+func (pl prometheusFollower) query() {
+	httpResp, err := http.Get(pl.server + "/api/v1/query?query=" + pl.expression)
+	if err != nil {
+		pl.errs <- fmt.Errorf("failed to fetch %q from prometheus server: %v", pl.expression, err)
+		return
+	}
+	defer httpResp.Body.Close()
+	type Response struct {
+		Status string `json:"status"`
+		Data   struct {
+			ResultType string `json:"resultType"`
+			Result     []struct {
+				Value []interface{} `json:"value"`
+			} `json:"result"`
+		} `data:"data"`
+	}
+
+	decoder := json.NewDecoder(httpResp.Body)
+	var resp Response
+	if err := decoder.Decode(&resp); err != nil {
+		pl.errs <- fmt.Errorf("failed to read response body from prometheus request for %q: %v", pl.expression, err)
+		return
+	}
+	fmt.Printf("xx: decoded json: %+v\n", resp)
+	if resp.Status != "success" {
+		pl.errs <- fmt.Errorf("prometheus request for %q returned non-success: %v", pl.expression, resp)
+		return
+	}
+	if len(resp.Data.Result) > 1 {
+		pl.errs <- fmt.Errorf("unexpectedly got %d results from prometheus request for %q: %+v", len(resp.Data.Result), pl.expression, resp)
+	}
+	firstResult := resp.Data.Result[0]
+	if len(firstResult.Value) != 2 {
+		// note: timestamp and value
+		pl.errs <- fmt.Errorf("unexpectedly got %d values from prometheus request for %q: %+v", len(firstResult.Value), pl.expression, resp)
+	}
+	timestamp := firstResult.Value[0]
+	value := firstResult.Value[1]
+	fmt.Printf("xx: parsed value %+v (%T) at time %v (%T)\n", value, value, timestamp, timestamp)
```

Comment:
what's the distinction between logging immediately vs. putting it into the logs channel?

Reply:
The `xx` messages are debug stuff so that it's easier to get a sense for the flow of logic for those that will maintain the project once I'm not actively contributing to it, including demo'ing how `%T` can log a variable's type etc. I'd expect that once this tool does something useful overall (see my other comment), it'd all be structured messages that go into the channels, with maybe some minimal logging in the main goroutine.